File: servers/cu_rs/ao-common/src/arweave.rs
use arweave_rs::{Arweave, ArweaveBuilder};
use futures::{Future, FutureExt};
use reqwest::{Response, Client, Error, Url};
use serde::{Deserialize, Serialize};
use std::{path::PathBuf, pin::Pin};
#[allow(unused)]
use log::{error, info};
use crate::errors::QueryGatewayErrors;
use crate::network::utils::get_content_type_headers;
use crate::models::gql_models::{Node, TransactionConnectionSchema};
use crate::models::shared_models::Tag;

pub struct InternalArweave {
    internal_arweave: Arweave,
    client: Client
}

impl InternalArweave {
    pub fn new(keypair_path: &str) -> Self {
        InternalArweave {
            internal_arweave: InternalArweave::create_wallet_client(keypair_path),
            client: Client::new()
        }
    }

    pub fn create_wallet_client(keypair_path: &str) -> Arweave {
        let arweave_builder = ArweaveBuilder::new();
        
        let mut path = PathBuf::new();
        path.push(keypair_path);
        arweave_builder
            .keypair_path(path)
            .base_url(Url::parse("https://arweave.net").unwrap())
            .build().unwrap()
        // arweave.upload_file_from_path(file_path, additional_tags, fee);
    }

    pub fn address_with<'a>(&'a self) -> Box<dyn Fn() -> String + 'a> {
        Box::new(|| self.internal_arweave.get_wallet_address().unwrap().clone())
    }

    pub fn build_sign_dataitem_with(&self) -> impl FnOnce(Vec<u8>, Vec<Tag>, String) -> Pin<Box<dyn Future<Output = DataItem>>> {
        move |data: Vec<u8>, tags: Vec<Tag>, anchor: String| {
            async move {
                DataItem {
                    id: "".to_string(),
                    data,
                    tags,
                    anchor            
                }
            }.boxed()
        }
    }

    /**
     * @typedef Env1
     * @property {fetch} fetch
     * @property {string} GATEWAY_URL
     *
     * @callback LoadTransactionMeta
     * @param {string} id - the id of the process whose src is being loaded
     * @returns {Async<z.infer<typeof transactionConnectionSchema>['data']['transactions']['edges'][number]['node']>}
     *
     * @param {Env1} env
     * @returns {LoadTransactionMeta}
    */
    pub async fn load_tx_meta_with(self, gateway_url: &str, id: &str) -> Result<Node, QueryGatewayErrors> {
        #[allow(non_snake_case)]
        let GET_PROCESSES_QUERY = r#"
            query GetProcesses ($processIds: [ID!]!) {
            transactions(ids: $processIds) {
                edges {
                node {
                    id
                    signature
                    anchor
                    owner {
                    address
                    }
                    tags {
                    name
                    value
                    }
                }
                }
            }
        }"#;

        let result = self.query_gateway_with::<ProcessIds, TransactionConnectionSchema>(gateway_url, GET_PROCESSES_QUERY, ProcessIds {
            process_ids: vec![id.to_string()]
        }).await;

        match result {
            Ok(tx) => {
                Ok(tx.data.transactions.edges[0].node.clone())
            },
            Err(e) => {
                error!("Error Encountered when fetching transaction {} from gateway {}", id, gateway_url);
                Err(e)
            }
        }
    }

    /**
    * @typedef Env2
    * @property {fetch} fetch
    * @property {string} GATEWAY_URL
    *
    * @callback LoadTransactionData
    * @param {string} id - the id of the process whose src is being loaded
    * @returns {Async<Response>}
    *
    * @param {Env2} env
    * @returns {LoadTransactionData}
    */
    pub async fn load_tx_data_with<'a>(&'a self, gateway_url: &'a str, id: &'a str) -> Result<Response, Error> {
        let result = self.client.get(format!("{}/raw/{}", gateway_url, id)).send().await;
        match result {
            Ok(res) => Ok(res),
            Err(e) => {
                error!("Error Encountered when fetching raw data for transaction {} from gateway {}", id, gateway_url);
                Err(e)
            }
        }
    }

    pub async fn query_gateway_with<'a, T: Serialize, U: for<'de> Deserialize<'de>>(&'a self, gateway_url: &'a str, query: &'a str, variables: T) -> 
        Result<U, QueryGatewayErrors> {        
        let result = self.client.post(format!("{}{}", gateway_url, "/graphql"))
            .headers(get_content_type_headers())
            .body(
                serde_json::to_string(&GraphqlQuery {
                    query: query.to_string(),
                    variables
                }).unwrap()
            )
            .send()
            .await;

        match result {
            Ok(res) => {
                let body_str = res.text().await.unwrap();
                match serde_json::from_str::<U>(&body_str) {
                    Ok(res) => Ok(res),
                    Err(e) => {                        
                        error!("Serialization error {:?}", e);
                        Err(QueryGatewayErrors::Serialization(Some(Box::new(e))))
                    }
                }
            },
            Err(e) => {
                error!("Error Encountered when querying gateway");
                Err(QueryGatewayErrors::Network(Some(Box::new(e))))
            }
        }
    }

    pub async fn upload_data_item_with<'a, U: for<'de> Deserialize<'de>>(&'a self, gateway_url: String, data_item: &DataItem) -> Result<U, Error> {
        let result = self.client
            .post(format!("{}/tx/arweave", gateway_url))
            .headers(get_content_type_headers())
            .body(serde_json::to_string(data_item).unwrap())
            .send()
            .await;

        match result {
            Ok(res) => Ok(res.json::<U>().await.unwrap()),
            Err(e) => {
                error!("Error while communicating with uploader:");
                Err(e)
            }
        }
    }
}

#[derive(Serialize)]
pub struct DataItem {
    pub id: String,
    pub data: Vec<u8>,
    pub tags: Vec<Tag>,
    pub anchor: String   
}

#[derive(Serialize)]
struct GraphqlQuery<T> {
    query: String,
    variables: T
}

/// variables type
#[derive(Serialize)]
struct ProcessIds {
    #[serde(rename = "processIds")]
    process_ids: Vec<String>
}

File: servers/cu/src/effects/arweave.js
import { Rejected, fromPromise, of } from 'hyper-async'
import { identity, path } from 'ramda'
import { z } from 'zod'
import Arweave from 'arweave'
import WarpArBundles from 'warp-arbundles'

import { backoff, joinUrl, okRes, okResWithNodeWith } from '../domain/utils.js'

const { createData, ArweaveSigner } = WarpArBundles

/**
 * @type {Arweave}
 */
let internalArweaveClient
export function createWalletClient () {
  if (internalArweaveClient) return internalArweaveClient
  internalArweaveClient = Arweave.init()
  return internalArweaveClient
}

export function addressWith ({ WALLET, arweave = internalArweaveClient }) {
  const wallet = JSON.parse(WALLET)
  const addressP = arweave.wallets.jwkToAddress(wallet)

  return () => addressP
}

export function buildAndSignDataItemWith ({ WALLET, createDataItem = createData }) {
  const signer = new ArweaveSigner(JSON.parse(WALLET))

  return async ({ data, tags, anchor }) => {
    const dataItem = createDataItem(data, signer, { anchor, tags })
    await dataItem.sign(signer)
    return {
      id: await dataItem.id,
      data: dataItem.getRaw()
    }
  }
}

/**
 * @typedef Env1
 * @property {fetch} fetch
 * @property {string} GRAPHQL_URL
 *
 * @callback LoadTransactionMeta
 * @param {string} id - the id of the process whose src is being loaded
 * @returns {Async<z.infer<typeof transactionConnectionSchema>['data']['transactions']['edges'][number]['node']>}
 *
 * @param {Env1} env
 * @returns {LoadTransactionMeta}
 */
export function loadTransactionMetaWith ({ fetch, GRAPHQL_URL, logger }) {
  // TODO: create a dataloader and use that to batch load contracts

  const GET_PROCESSES_QUERY = `
    query GetProcesses (
      $processIds: [ID!]!
      $skipTags: Boolean!
      $skipSignature: Boolean!
      $skipAnchor: Boolean!
    ) {
      transactions(ids: $processIds) {
        edges {
          node {
            id
            signature @skip (if: $skipSignature)
            anchor @skip (if: $skipAnchor)
            owner {
              address
              key
            }
            tags @skip (if: $skipTags) {
              name
              value
            }
            recipient
          }
        }
      }
    }`

  const transactionConnectionSchema = z.object({
    data: z.object({
      transactions: z.object({
        edges: z.array(z.object({
          node: z.record(z.any())
        }))
      })
    })
  })

  /**
   * Ideally, this would be in the function schema,
   * but there is a bug with zod and optional function\
   * args https://github.com/colinhacks/zod/issues/2990
   *
   * So we manually run it
   */
  const optionsSchema = z.object({
    skipTags: z.boolean().default(false),
    skipSignature: z.boolean().default(false),
    skipAnchor: z.boolean().default(false)
  }).default({})

  return (id, options) => {
    const okResWithNode = okResWithNodeWith(logger, id)
    return of(id)
      .map((id) => {
        const variables = optionsSchema.parse(options)
        variables.processIds = [id]
        return variables
      })
      .chain(fromPromise((variables) =>
        backoff(
          () => fetch(GRAPHQL_URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ query: GET_PROCESSES_QUERY, variables })
          }).then(okResWithNode),
          { maxRetries: 5, delay: 500, log: logger, name: `loadTransactionMeta(${JSON.stringify({ id })})` }
        )
          .then(transactionConnectionSchema.parse)
          .then(path(['data', 'transactions', 'edges', '0', 'node']))
          .then((node) => {
            if (node) return node
            logger('Transaction "%s" was not found on gateway', id)
            // TODO: better error handling
            const err = new Error(`Transaction '${id}' not found on gateway`)
            err.status = 404
            throw err
          })
      ))
      .toPromise()
  }
}

/**
   * @typedef Env2
   * @property {fetch} fetch
   * @property {string} ARWEAVE_URL
   *
   * @callback LoadTransactionData
   * @param {string} id - the id of the process whose src is being loaded
   * @returns {Async<Response>}
   *
   * @param {Env2} env
   * @returns {LoadTransactionData}
   */
export function loadTransactionDataWith ({ fetch, ARWEAVE_URL, logger }) {
  // TODO: create a dataloader and use that to batch load processes
  return (id) =>
    of(id)
      .chain(fromPromise((id) =>
        backoff(
          () => fetch(joinUrl({ url: ARWEAVE_URL, path: `/raw/${id}` })).then(okRes),
          { maxRetries: 5, delay: 500, log: logger, name: `loadTransactionData(${JSON.stringify({ id })})` }
        )
          .then(async (res) => {
            if (res.ok) return res
            logger('Error Encountered when fetching raw data for transaction \'%s\'', id)
            throw new Error(`Error encountered when fetching transaction '${id}' from arweave: ${res.status}: ${await res.text()}`)
          })
      ))
      .toPromise()
}

export function queryGatewayWith ({ fetch, GRAPHQL_URL, logger }) {
  return async ({ query, variables }) => {
    return backoff(
      () =>
        fetch(GRAPHQL_URL, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ query, variables })
        }).then(okRes),
      { maxRetries: 5, delay: 500, log: logger, name: `queryGateway(${JSON.stringify({ query, variables })})` }
    )
      .then(async (res) => {
        if (res.ok) return res.json()
        logger('Error Encountered when querying gateway')
        throw new Error(`${res.status}: ${await res.text()}`)
      })
  }
}

export function uploadDataItemWith ({ UPLOADER_URL, fetch, logger }) {
  return async (dataItem) => {
    return of(dataItem)
      .chain(fromPromise((body) =>
        fetch(`${UPLOADER_URL}/tx/arweave`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/octet-stream',
            Accept: 'application/json'
          },
          body
        })
      ))
      .bimap(
        logger.tap('Error while communicating with uploader: %O'),
        identity
      )
      .bichain(
        (err) => Rejected(JSON.stringify(err)),
        fromPromise(async (res) => {
          if (!res?.ok) {
            const text = await res.text()
            throw new Error(`${res.status}: ${text}`)
          }
          return res.json()
        })
      )
      .toPromise()
  }
}
